quinn-proto/src/connection/streams/send.rs

```diff
@@ -66,12 +69,12 @@ impl Send {
         }
         let limit = limit.min(budget) as usize;
 
-        let mut chunks = Vec::new();
-        let source_output = source(limit, &mut chunks);
+        debug_assert!(self.chunks.is_empty());
+        let source_output = source(limit, &mut self.chunks);
 
         let mut written = 0;
 
-        for mut chunk in chunks.drain(..) {
+        for mut chunk in self.chunks.drain(..) {
             let prefix = chunk.split_to(chunk.len().min(limit - written));
```

Comment:
It might be nice to skip the `split_to` when we can consume the whole chunk. It's at least an indirect call, and might conceivably fiddle with refcounts.

Reply:
It looks like `split_to` already has an optimization for this, so I don't think we need to :)

https://docs.rs/bytes/latest/src/bytes/bytes.rs.html#548-571

```rs
        if at == self.len() {
            let end_ptr = self.ptr.wrapping_add(at);
            return mem::replace(self, Bytes::new_empty_with_ptr(end_ptr));
        }
```